### DISCUSSION TITLE
Speaker embedding layer already initialized before d_vector settings

### DISCUSSION THREAD

**Discussion by Ca-ressemble-a-du-fake** [coqui-ai/TTS#1164] [General Q&A]
Hi,

While fine-tuning the VITS model from Experiment 3 En-Pt-Fr + SCL checkpoint, I get the following error :
`Speaker embedding layer already initialized before d_vector settings`. 
The error stack reads : 
`
> initialization of speaker-embedding layers.
Traceback (most recent call last):
  File "/content/drive/MyDrive/YourTTS/TTS/configs/vits_ljspeech.py", line 221, in <module>
    model = Vits(config)
  File "/content/TTS/TTS/tts/models/vits.py", line 303, in __init__
    self.init_multispeaker(config)
  File "/content/TTS/TTS/tts/models/vits.py", line 404, in init_multispeaker
    self._init_d_vector()
  File "/content/TTS/TTS/tts/models/vits.py", line 439, in _init_d_vector
    raise ValueError("[!] Speaker embedding layer already initialized before d_vector settings.")
ValueError: [!] Speaker embedding layer already initialized before d_vector settings.

`
I tried with and without `use_speaker_embedding=True` but the error is the same. What can I do to overcome this issue ?

THanks you

**Comment by jaggukaka** [comment]
Facing similar issue, while fine tuning, if both use_speaker_embedding and use_d_vector_file set to true, then the model training halting with the above mentioned error.

**Comment by vince62s** [comment]
https://github.com/coqui-ai/TTS/discussions/1058#discussioncomment-1939403

**Reply by Ca-ressemble-a-du-fake** [reply]
Thanks @vince62s, I am trying to generate the d_vector with the provided script. How do you create the dataset_config json , per hand ? I could not find examples.

@Edresson wrote in a [comment](https://gitter.im/coqui-ai/TTS?at=61d84df6d41a5853f93b4796) to use `config.save_json(config_json_path)` and I put it in my `train_vits.py`.

**Reply by jaggukaka** [reply]
Just use something like this

```
{   
    "model": "speaker_encoder",
    "run_name": "speaker_encoder",
    "run_description": "resnet speaker encoder trained with commonvoice all languages dev and train, Voxceleb 1 dev and Voxceleb 2 dev",
    "datasets": [
        {
            "name": "<your_formatter_name>",
            "path": "<path_to_dataset>",
            "meta_file_train": "<path_to_meta_file>",
            "ignored_speakers": null,
            "language": "en",
            "meta_file_val": null,
            "meta_file_attn_mask": ""
        }
    ]
}
```

**Reply by Ca-ressemble-a-du-fake** [reply]
Thanks a lot @jaggukaka , that helped, I could launch the embedding generation with your example!

**Comment by jaggukaka** [comment]
> 

Just for completion.
If you are using d_vectors then make sure following properties are set in the config file (Still need to verify this...)

```
{
"use_d_vector_file" : true,
"use_speaker_embedding" : false,
"speaker_encoder_config_path" : "<path_to_speakerencoder_config>"
"speaker_encoder_model_path" : "<path_to_speakerencoder_model>"
"use_speaker_encoder_as_loss" : true
}
```

If you don't want to use d_vectors and instead want the yourtts model to have a speaker_embedding layer which is trained along with the model, make sure following properties are set in the config file (I tried only this.)

```
{
"use_d_vector_file" : false,
"use_speaker_embedding" : true,
"use_speaker_encoder_as_loss" : false
}
```

**Reply by Ca-ressemble-a-du-fake** [reply]
Thank you @jaggukaka . I tried your second suggestion but got 
`
 >  initialization of speaker-embedding layers.
 > Using CUDA:  True
 > Number of GPUs:  1
Traceback (most recent call last):
  File "/content/drive/MyDrive/YourTTS/TTS/configs/vits_ljspeech.py", line 258, in <module>
    training_assets={"audio_processor": ap},
  File "/content/TTS/TTS/trainer.py", line 284, in __init__
    self.optimizer = self.get_optimizer(self.model, self.config)
  File "/content/TTS/TTS/trainer.py", line 1068, in get_optimizer
    optimizer = model.get_optimizer()
  File "/content/TTS/TTS/tts/models/vits.py", line 947, in get_optimizer
    gen_parameters = chain(gen_parameters, self.emb_l.parameters())
AttributeError: 'NoneType' object has no attribute 'parameters'

`
Which recipe did you use ? Because I read somewhere that `the recipe under multilingual/vits isn't the one used for YourTTS. It doesn't use d_vectors but internal embeddings.`

**Reply by jaggukaka** [reply]
@Ca-ressemble-a-du-fake I didn't use any recipe, I just used my custom config.json and train_tts.py script. 
I got the d_vectors way working only with this fix which I did in my [repo](https://github.com/jaggukaka/TTS/commit/86df90eebf8969f4dbd84f9b12b9bd8639c9efbf#diff-0eee2f633e8e50395b61a21049db8421db45c06b2205595fe7968eb5ad4b1a1fL433)

**Reply by Ca-ressemble-a-du-fake** [reply]
@jaggukaka thank you, I think I am using the recipe the wrong way. I will try with custom config.json as you did.
